source/module_hamilt_pw/hamilt_pwdft/fs_nonlocal_tools.cpp

```diff
@@ -0,0 +1,1077 @@
+                                     this->nlpp_->nhtol.nc,
+                                     npw,
+                                     it,
+                                     ipol,
+                                     jpol,
+                                     vkb_deri_ptr,
+                                     vkb_ptrs,
+                                     hd_ylm,
+                                     ylm_ptrs,
+                                     hd_ylm_deri,
+                                     ylm_deri_ptrs1,
+                                     ylm_deri_ptrs2,
+                                     hd_vq,
+                                     vq_ptrs,
+                                     hd_vq_deri,
+                                     vq_deri_ptrs);
+                cal_vkb_deri_op()(this->ctx,
+                                  nh,
+                                  npw,
+                                  ipol,
+                                  jpol,
+                                  vq_ptrs,
+                                  vq_deri_ptrs,
+                                  ylm_ptrs,
+                                  ylm_deri_ptrs1,
+                                  ylm_deri_ptrs2,
+                                  sk,
+                                  pref.data(),
+                                  g_plus_k.data(),
+                                  vkb_ptrs);
+            }
+            delete[] sk;
+
+            vkb_deri_ptr += nh * npw;
+        }
+    }
+    // 2.b calculate dbecp = dbecp_noevc * psi
+    const char transa = 'C';
+    const char transb = 'N';
+
+    gemm_op()(this->ctx,
+              transa,
+              transb,
+              nkb,
+              npm,
+              npw,
+              &ModuleBase::ONE,
+              ppcell_vkb,
+              npw,
+              ppsi,
+              this->max_npw,
+              &ModuleBase::ZERO,
+              dbecp,
+              nkb);
+    // calculate stress for target (ipol, jpol)
+    const int current_spin = this->kv_->isk[ik];
+    cal_stress_nl_op()(this->ctx,
+                       nondiagonal,
+                       ipol,
+                       jpol,
+                       nkb,
+                       npm,
+                       this->ntype,
+                       current_spin, // uspp only
+                       this->nbands,
+                       ik,
+                       this->nlpp_->deeq.getBound2(),
+                       this->nlpp_->deeq.getBound3(),
+                       this->nlpp_->deeq.getBound4(),
+                       atom_nh,
+                       atom_na,
+                       d_wg,
+                       d_ekb,
+                       qq_nt,
+                       deeq,
+                       becp,
+                       dbecp,
+                       stress);
+}
+
+template <typename FPTYPE, typename Device>
+void FS_Nonlocal_tools<FPTYPE, Device>::cal_dbecp_f(int ik, int npm, int ipol)
```

Comment:
In terms of safety, `f(const int x);` is regarded as safer in the context that it protects the parameter from unintended modifications within the function. This can prevent accidental bugs where a function inadvertently changes the value of a parameter when it wasn't meant to. Although, for a by-value parameter, where the argument passed in is not affected by changes made within the function, the practical benefit is more about clarity of intention rather than preventing bugs that would alter the caller's variable state.

I think the advise are good, and maybe we can use robot to modify them automaticly.